cpp/doc/libcuspatial_refactoring_guide.md

```diff
@@ -0,0 +1,272 @@
+
+  1. Header-only APIs can increase compilation time for code that depends on them.
+  2. Some users (especially our Python API) may prefer a cuDF-based API.
+
+The good news is that by maintaining the existing libcudf-based C++ API as a layer above the header-
+only libcuspatial API, we can avoid problem 1 and problem 2 for users of the legacy API.
+
+## Example API
+
+Following is an example iterator-based API for `cuspatial::haversine_distance`.
+
+```C++
+template <class LonLatItA,
+          class LonLatItB,
+          class OutputIt,
+          class Location = typename std::iterator_traits<LonLatItA>::value_type,
+          class T        = typename Location::value_type>
+OutputIt haversine_distance(LonLatItA a_lonlat_first,
+                            LonLatItA a_lonlat_last,
+                            LonLatItB b_lonlat_first,
+                            OutputIt distance_first,
+                            T const radius = EARTH_RADIUS_KM,
+                            rmm::cuda_stream_view stream = rmm::cuda_stream_default);
+```
+
+There are a few key points to notice.
+
+  1. The API is very similar to STL algorithms such as `std::transform`.
+  2. All array inputs and outputs are iterator type templates. 
+  3. Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`
+     type (include/cuspatial/types.hpp)
+  4. The `Location` type is a template that is by default equal to the `value_type` of the input
+     iterators.
+  5. The floating point type is a template (`T`) that is by default equal to the `value_type` of
+     `Location`.
+  6. The iterator types for the two input ranges (A and B) are distinct templates. This is crucial
+     to enable composition of fancy iterators that may be different types for A and B.
+  7. The size of the input and output ranges in the example API are equal, so the start and end of
+     only the A range is provided (`a_lonlat_first` and `a_lonlat_last`). This mirrors STL APIs.
+  8. This API returns an iterator to the element past the last element written to the output. This
+     is inspired by `std::transform`, even though as with `transform`, many uses of 
+     `haversine_distance` will not need this returned iterator.
+  9. All APIs that run CUDA device code (including Thrust algorithms) or allocate memory take a CUDA
+     stream on which to execute the device code and allocate memory.
+
+## Example Documentation
+
+Following is the (Doxygen) documentation for the above `cuspatial::haversine_distance`.
+
+```C++
+/**
+ * @brief Compute haversine distances between points in set A to the corresponding points in set B.
+ *
+ * Computes N haversine distances, where N is `std::distance(a_lon_first, a_lon_last)`.
+ * The distance for each `(a_lon[i], a_lat[i])` and `(b_lon[i], b_lat[i])` point pair is assigned to
+ * `distance_first[i]`. `distance_first` must be an iterator to output storage allocated for N
+ * distances.
```

Comment:
I find it nice when documentation explicitly calls out if it's okay for any of the iterator ranges to overlap. On the other hand, you could argue that unless it explicitly says that they cannot then it implies that they can.

E.g., `thrust::scatter` references this in the "Precondition" section: https://thrust.github.io/doc/group__scattering_gac1b33a02dd67cf9c8b3b3aa05c881a37.html

(You may want to consider having an explicit "Precondition" section. I've been leaning more and more in this direction for documentation.)

Reply:
Done. Used @pre and added preconditions regarding overlapping ranges, modeled after the Thrust docs.